### PR TITLE
TLA+: Forking cases for tendermint

### PR DESCRIPTION
**Pull request by konnov**
This is the PR moved from the [verification repository](https://github.com/informalsystems/verification/pull/13).

It contains a TLA+ specification of Tendermint consensus that is tuned to safety and fork scenarios.

### CODE REVIEW

File: docs/spec/tendermint-fork-cases/TendermintAcc3.tla [hunk 1]
```diff
@@ -0,0 +1,447 @@
...
+THRESHOLD2 == 2 * T + 1 \* a quorum when having N > 3 * T
+
+(********************* TYPE ANNOTATIONS FOR APALACHE **************************)
+\* the operator for type annotations
+a <: b == a
+
+\* the type of message records
+MT == [type |-> STRING, src |-> STRING, round |-> Int,
+       proposal |-> STRING, validRound |-> Int, id |-> STRING]
+
```
Comment (istoilkovska) [inline docs/spec/tendermint-fork-cases/TendermintAcc3.tla:87]: Some comments on what the message record fields encode would be useful. For example, it is not clear that `src` is a process ID or that proposal is one of `"PROPOSAL", "PREVOTE", "PRECOMMIT"`. Not sure if this is the appropriate place, maybe they can go in the file where the invariants are defined.

File: docs/spec/tendermint-fork-cases/TendermintAcc3.tla [hunk 2]
```diff
@@ -0,0 +1,447 @@
...
+    AsMsg([type |-> "PROPOSAL", src |-> pSrc, round |-> pRound,
+           proposal |-> pProposal, validRound |-> pValidRound])
+  IN
+  msgsPropose' = [msgsPropose EXCEPT ![pRound] = msgsPropose[pRound] \cup {newMsg}]
+
+BroadcastPrevote(pSrc, pRound, pId) ==
+  LET newMsg == AsMsg([type |-> "PREVOTE",
+                       src |-> pSrc, round |-> pRound, id |-> pId])
+  IN
+  msgsPrevote' = [msgsPrevote EXCEPT ![pRound] = msgsPrevote[pRound] \cup {newMsg}]
```
Comment (istoilkovska) [inline docs/spec/tendermint-fork-cases/TendermintAcc3.tla:174]:
```suggestion
  msgsPrevote' = [msgsPrevote EXCEPT ![pRound] = msgsPrevote[pRound] \union {newMsg}]
```

File: docs/spec/tendermint-fork-cases/TendermintAcc3.tla [hunk 3]
```diff
@@ -0,0 +1,447 @@
...
+
+BroadcastPrecommit(pSrc, pRound, pId) ==
+  LET newMsg == AsMsg([type |-> "PRECOMMIT",
+                       src |-> pSrc, round |-> pRound, id |-> pId])
+  IN
+  msgsPrecommit' = [msgsPrecommit EXCEPT ![pRound] = msgsPrecommit[pRound] \cup {newMsg}]
```
Comment (istoilkovska) [inline docs/spec/tendermint-fork-cases/TendermintAcc3.tla:180]:
```suggestion
  msgsPrecommit' = [msgsPrecommit EXCEPT ![pRound] = msgsPrecommit[pRound] \union {newMsg}]
```

File: docs/spec/tendermint-fork-cases/MC_n10_f3.tla [hunk 4]
```diff
@@ -0,0 +1,26 @@
+----------------------------- MODULE MC_n10_f3 -------------------------------
```
Comment (istoilkovska) [inline docs/spec/tendermint-fork-cases/MC_n10_f3.tla:1]: Are these MC files generated automatically, or does one have to do copy-paste?

File: docs/spec/tendermint-fork-cases/README.md [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+Third, in the general case, we either have no fork, or two fork scenarios:
+
+```tla
+THEOREM AgreementOrFork ==
+    ~FaultyQuorum /\ TypedInv => AgreementOrEquivocationOrAmnesia
+```
+
+# Model checking results   
+
+Check the report on [model checking with Apalache](./results/001indinv-apalache-report.md).
```
Comment (istoilkovska) [inline docs/spec/tendermint-fork-cases/README.md:88]: The .md file does not exist.

File: docs/spec/tendermint-fork-cases/TendermintAcc3.tla [hunk 6]
```diff
@@ -0,0 +1,447 @@
...
+MT == [type |-> STRING, src |-> STRING, round |-> Int,
+       proposal |-> STRING, validRound |-> Int, id |-> STRING]
+       
+\* a type annotation for a message
+AsMsg(m) == m <: MT
+\* a type annotation for a set of messages
+SetOfMsgs(S) == S <: {MT}       
+\* a type annotation for an empty set of messages
+EmptyMsgSet == SetOfMsgs({})
+
```
[the remark this answers is not in the document]
Comment (konnov) [reply docs/spec/tendermint-fork-cases/TendermintAcc3.tla:94]: We are now refactoring type annotations and type inference, so they will look completely different but much more like in programming languages: https://github.com/informalsystems/apalache/blob/unstable/docs/adr/002adr-types.md

These types will be an integral part of a specification, not ad-hoc type annotations.

File: docs/spec/tendermint-fork-cases/MC_n10_f3.tla (hunk 4 above)
Comment (konnov) [reply docs/spec/tendermint-fork-cases/MC_n10_f3.tla:1]: These ones a produced by copy-paste :(

File: docs/spec/tendermint-fork-cases/README.md (hunk 5 above)
Comment (konnov) [reply docs/spec/tendermint-fork-cases/README.md:88]: Yeah. I have added the missing files.

**Review by josef-widder — APPROVED** [review]
lgtm. Perhaps we will revisit an polish in the future.

File: docs/spec/tendermint-fork-cases/TendermintAcc3.tla (hunk 1 above)
Comment (konnov) [reply docs/spec/tendermint-fork-cases/TendermintAcc3.tla:87]: Right. This will change with the new type checker anyways.

**Pull request merged**